aya/src/maps/lpm_trie.rs

```diff
@@ -0,0 +1,456 @@
+    /// represents the number of bytes matched against.
+    pub prefix_len: u32,
+    /// represents arbitrary data stored in the LPMTrie.
+    pub data: K,
+}
+
+impl<K: Pod> Key<K> {
+    /// A constructor for the Key struct.
+    /// ```no_run
+    /// use aya::maps::lpm_trie::{LpmTrie, Key};
+    /// use std::convert::TryFrom;
+    /// use std::net::Ipv4Addr;
+    ///
+    /// let ipaddr = Ipv4Addr::new(8,8,8,8);
+    /// let key =  Key::new(16, u32::from(ipaddr).to_be());
+    /// ```
+    pub fn new(prefix_len: u32, data: K) -> Self {
+        Self { prefix_len, data }
+    }
+}
+
+impl<K: Pod> Copy for Key<K> {}
+
+impl<K: Pod> Clone for Key<K> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+// A Pod impl is required as Key struct is a key for a map.
+unsafe impl<K: Pod> Pod for Key<K> {}
+
+impl<T: Deref<Target = Map>, K: Pod, V: Pod> LpmTrie<T, K, V> {
+    pub(crate) fn new(map: T) -> Result<LpmTrie<T, K, V>, MapError> {
+        let map_type = map.obj.def.map_type;
+
+        // validate the map definition
+        if map_type != BPF_MAP_TYPE_LPM_TRIE as u32 {
+            return Err(MapError::InvalidMapType {
+                map_type: map_type as u32,
+            });
+        }
+        let size = mem::size_of::<Key<K>>();
+        let expected = map.obj.def.key_size as usize;
+        if size != expected {
+            return Err(MapError::InvalidKeySize { size, expected });
+        }
+        let size = mem::size_of::<V>();
+        let expected = map.obj.def.value_size as usize;
+        if size != expected {
+            return Err(MapError::InvalidValueSize { size, expected });
+        };
+
+        let _ = map.fd_or_err()?;
+
+        Ok(LpmTrie {
+            inner: map,
+            _k: PhantomData,
+            _v: PhantomData,
+        })
+    }
+
+    /// Returns a copy of the value associated with longest-prefix-matching key in the LPMTrie.
+    pub fn get(&self, key: &Key<K>, flags: u64) -> Result<V, MapError> {
+        let fd = self.inner.deref().fd_or_err()?;
+        let value = bpf_map_lookup_elem(fd, key, flags).map_err(|(code, io_error)| {
+            MapError::SyscallError {
+                call: "bpf_map_lookup_elem".to_owned(),
+                code,
+                io_error,
+            }
+        })?;
+        value.ok_or(MapError::KeyNotFound)
+    }
+
+    /// Inserts a key-value pair for the value associated with the key.
+    pub fn insert(&self, key: &Key<K>, value: V, flags: u64) -> Result<(), MapError> {
+        let fd = self.inner.deref().fd_or_err()?;
+        bpf_map_update_elem(fd, key, &value, flags).map_err(|(code, io_error)| {
+            MapError::SyscallError {
+                call: "bpf_map_update_elem".to_owned(),
+                code,
+                io_error,
+            }
+        })?;
+
+        Ok(())
+    }
+
+    /// Deletes elements from the map by key.
+    /// This removes elements with an *exact* match - both prefix_len and data.
```

Comment:
So can this actually remove multiple nodes? Can you have multiple nodes with the same prefix len and data?

Reply:
No - read further :) only an *exact* match would be removed.

Reply:
>Can you have multiple nodes with the same prefix len and data?

That's just assignment to the same key (same as with any map)

Reply:
Then the doc shouldn't use the plural form, it should say:

```
Removes an element from the map.

Both the prefix and data must match exactly - this method does not do a longest prefix match.
```

Reply:
Ah! thanks a lot! great catch.

Reply:
Done.